File: 23.py
def maxSubarraySumCircular(nums):
    total_sum = sum(nums)
    
    # Kadane’s for max subarray sum (non-circular)
    cur_max = max_sum = nums[0]
    # Kadane’s for min subarray sum
    cur_min = min_sum = nums[0]
    
    for n in nums[1:]:
        cur_max = max(n, cur_max + n)
        max_sum = max(max_sum, cur_max)

        cur_min = min(n, cur_min + n)
        min_sum = min(min_sum, cur_min)
    
    # Case 1: non-circular max_sum
    # Case 2: circular -> total_sum - min_sum
    if max_sum < 0:
        return max_sum  # All numbers are negative
    return max(max_sum, total_sum - min_sum)


# Example
print(maxSubarraySumCircular([1, -2, 3, -2]))  # Output: 3

File: 23.js
function maxSubarraySumCircular(nums) {
  const total = nums.reduce((a, b) => a + b, 0);

  let curMax = nums[0], maxSum = nums[0];
  let curMin = nums[0], minSum = nums[0];

  for (let i = 1; i < nums.length; i++) {
    curMax = Math.max(nums[i], curMax + nums[i]);
    maxSum = Math.max(maxSum, curMax);

    curMin = Math.min(nums[i], curMin + nums[i]);
    minSum = Math.min(minSum, curMin);
  }

  // If all are negative, return the maximum single element
  if (maxSum < 0) return maxSum;
  
  return Math.max(maxSum, total - minSum);
}

// Example
console.log(maxSubarraySumCircular([1, -2, 3, -2])); // Output: 3
